drivers/wifi/simplelink.c

```diff
@@ -0,0 +1,183 @@
+/**
+ * Copyright (c) 2018 Texas Instruments, Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_WIFI_LEVEL
+#define SYS_LOG_DOMAIN "dev/simplelink"
+#include <logging/sys_log.h>
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <device.h>
+#include <net/net_if.h>
+#include <net/wifi_mgmt.h>
+#include <net/net_offload.h>
+
+#include "simplelink_support.h"
+
+struct simplelink_data {
+	struct net_if *iface;
+	unsigned char mac[6];
+};
+
+static struct simplelink_data simplelink_data;
+
+/* Handle connection events from the SimpleLink Event Handlers: */
+static void simplelink_wifi_cb(u32_t mgmt_event, struct sl_connect_state *conn)
+{
+	struct in_addr addr;
+	struct in_addr gwaddr;
+	int status;
+
+	/*
+	 * Once Zephyr wifi_mgmt wifi_status codes are defined, will need
+	 * to map from SimpleLink error codes.  For now, just return -EIO.
+	 */
+	status = (conn->error ? -EIO : 0);
+
+	switch (mgmt_event) {
+	case NET_EVENT_WIFI_CMD_CONNECT_RESULT:
+		/* Only get this event if connect succeeds: */
+		wifi_mgmt_raise_connect_result_event(simplelink_data.iface,
+						     status);
+		k_yield(); /* Needed to fire net_event notifications: */
+		break;
+
+	case NET_EVENT_WIFI_CMD_DISCONNECT_RESULT:
+		/* Could be during a connect, disconnect, or async error: */
+		wifi_mgmt_raise_disconnect_result_event(simplelink_data.iface,
+							status);
+		k_yield(); /* Needed to fire net_event notifications: */
+		break;
+
+	case NET_EVENT_IPV4_ADDR_ADD:
+		addr.s_addr = htonl(conn->ip_addr);
+		gwaddr.s_addr = htonl(conn->gateway_ip);
+		net_if_ipv4_set_gw(simplelink_data.iface, &gwaddr);
+		net_if_ipv4_addr_add(simplelink_data.iface, &addr,
+				     NET_ADDR_DHCP, 0);
+		break;
+
+	default:
+		SYS_LOG_DBG("Unrecognized mgmt event: 0x%x", mgmt_event);
+		break;
+	}
+}
+
+/* TBD: Only here to link/test WiFi mgmnt part */
+static struct net_offload simplelink_offload = {
+	.get            = NULL,
+	.bind		= NULL,
+	.listen		= NULL,
+	.connect	= NULL,
+	.accept		= NULL,
+	.send		= NULL,
+	.sendto		= NULL,
+	.recv		= NULL,
+	.put		= NULL,
+};
+
+static int simplelink_mgmt_scan(struct device *dev, scan_result_cb_t cb)
+{
+	int index = 0;
+	int num_results_or_err; /* error is negative */
+	int status;
+	struct wifi_scan_result scan_result;
+	struct net_if *iface =
+		((struct simplelink_data *)dev->driver_data)->iface;
+
+	/* Get the table of scan results: */
+	num_results_or_err = _simplelink_scan();
+	status = (num_results_or_err < 0 ? -EIO : 0);
+
+	/* Now iterate over the table, and call the scan_result callback. */
+	if (cb) {
```

Comment:
callback is always provided. no need to test it validity

Reply:
There is no contract in `wifi_mgmt.h:net_wifi_mgmt_offload{... scan() ...}` to specify that cb is always valid.
The current code in `wifi_mgmt.c:wifi_scan()` ensures cb is valid, until someone changes that code later on, causing wifi drivers' scan functions to crash.
So, I can change the check to an assert() to test this hidden assumption.

Reply:
yeah well there won't be numerous implementation of wifi_mgmt: only one. So what it does _is_ the contract (all of our APIs lacks of exhaustive documentation after all)